x/concentrated-liquidity/pool_hooks_test.go

```diff
@@ -178,9 +186,131 @@ func (s *KeeperTestSuite) TestCallPoolActionListener() {
 	}
 }
 
+// Pool hook tests
+// General testing strategy:
+//  1. Build a pre-defined contract that defines the following behavior for all hooks:
+//     if triggered, transfer 1 token with denom corresponding to the action prefix
+//     e.g. if action prefix is "beforeSwap", transfer 1 token with denom "beforeSwap"
+//  2. Set this contract for all hooks defined by the test case (each case should have a list
+//     of action prefixes it wants to "activate")
+//  3. Run a series of actions that would trigger all the hooks (create, withdraw from, swap against a position),
+//     and ensure that the correct denoms are in the account balance after each action/at the end.
+//
+// NOTE: we assume that set contracts have valid implementations for all hooks and that this is validated
+// at the contract setting stage at a higher level of abstraction. Thus, this class of errors is not covered
+// by these tests.
+func (s *KeeperTestSuite) TestPoolHooks() {
+	hookContractFilePath := "./testcontracts/compiled-wasm/hooks.wasm"
+
+	allBeforeHooks := []string{
+		before(types.CreatePositionPrefix),
+		before(types.WithdrawPositionPrefix),
+		before(types.SwapExactAmountInPrefix),
+		before(types.SwapExactAmountOutPrefix),
+	}
+
+	allAfterHooks := []string{
+		after(types.CreatePositionPrefix),
+		after(types.WithdrawPositionPrefix),
+		after(types.SwapExactAmountInPrefix),
+		after(types.SwapExactAmountOutPrefix),
+	}
+
+	allHooks := append(allBeforeHooks, allAfterHooks...)
+
+	testCases := map[string]struct {
+		actionPrefixes []string
+	}{
+		"single hook: before create position": {
+			actionPrefixes: []string{before(types.CreatePositionPrefix)},
+		},
+		"all before hooks": {
+			actionPrefixes: allBeforeHooks,
+		},
+		"all after hooks": {
+			actionPrefixes: allAfterHooks,
+		},
+		"all hooks": {
+			actionPrefixes: allHooks,
+		},
```

Comment:
Seems like there will be dupicated test runs for before hook & after hook, was it that we wanted to test compound integration of hooks? 👀

Reply:
Not sure I follow – the "all hooks" case is intended to run every hook simultaneously whereas the others run different subsets

Reply:
yep, I was just curious of the significance of running hooks simultaneously vs running subset. (But it was a nit question, can be ignored)

Reply:
Ah I see – no particular reason other than general coverage (ensuring that various subsets can correctly run without triggering the others)